### DISCUSSION TITLE
remove whitespace

### DISCUSSION THREAD

**Discussion by fkv1** [sqlalchemy/mako#391] [potential enhancements]
The Template Toolkit provides an [easy syntax](http://template-toolkit.org/docs/manual/Syntax.html#section_Chomping_Whitespace) for removing leading and trailing whitespace:
[%- instead of [% removes all whitespace (including newlines) before the directive
-%] instead of %] removes all whitespace (including newlines) after the directive

The latter is barely ever needed, but the removal of leading whitespace is very important.

This feature is missing in Mako. You can only add a backslash at the end of the line to eat the newline. But that's insufficient in several respects:

1. It's counter-intuitive. If you insert a block, you don't want to edit other lines to compensate for the effects of the inserted block. Same when you remove the block. There's a high probability you'll forget to remove the backslash in the preceding line.
2.  diffs and patches also become more confusing if the preceding line needs to be changed when you remove or insert a block.
3. if you have:
```
<div>
    aaa\
    <%
        (some python code)
    %>
    bbb
</div>
```
or
```
<div>
    aaa
    <%
        (some python code)
    %>\
    bbb
</div>
```
you get:
```
<div>
    aaa
        bbb
</div>
```
(because only the newline is chomped, not the spaces)

but you want:
```
<div>
    aaa
    bbb
</div>
```
Of course you can remove the indentation of the Mako block, but at the cost of readability.

The Template Toolkit approach would be:
```
<div>
    aaa
    <%-
        (some python code)
    %>
    bbb
</div>
```

Can this be implemented?

**Comment by zzzeek** [comment]
ive added a new category "potential enhancements" for discussions like this.

~~I am -1 on adding new syntaxes that replace the use of `<% %>`~~ edit, thought you were proposing `[% %]` with brackets.   not -1 on this syntax but still no resources to work on it.  

as mentioned elsewhere Mako hasnt had any syntax changes in over a decade and we aren't really in a place to be rethinking things.     The language is kind of brittle and even very small changes need lots of attention and testing and generally take a few years to not keep producing regressions and new issues.   I also am not implementing anything for Mako other than critical bug fixes if any these days and I definitely dont have the resources to maintain major new behaviors for months/years on my own.

all of that said, you can probably implement this syntax / feature yourself right now using a [preprocessor](https://docs.makotemplates.org/en/latest/usage.html#mako.template.Template.params.preprocessor) function that just replaces out `[% %]` blocks for `<% %>` with leading/trailing whitespace removed.

**Reply by zzzeek** [reply]
I wonder also if this idea has been discussed before, using dashes inside the `<% %>` is a fine place to do this sort of thing, maybe Jinja does this?
